### ISSUE TITLE
Using translations for encoding

### ISSUE THREAD

**Issue by tsloughter** [tomas-abrahamsson/gpb#178]
First I should explain what I'm trying to achieve in case it isn't what translations are meant for:

I have an [implementation](https://github.com/opentelemetry-beam/opentelemetry_zipkin) of the client for the Zipkin API which sends a proto:

```proto
message ListOfSpans {
  repeated Span spans = 1;
}
```

I looked into gpb translations hoping they were a way I could take my list of spans that are in an internal format to the tracer (https://github.com/open-telemetry/opentelemetry-erlang) and while constructing the `ListofSpans` proto by iterating over the list of spans convert them to the `Span` proto format.

Basically to not have to iterate over all the spans once converting to Zipkin Spans and then call `encode_msg(#list_of_spans{spans=ZipkinSpans})` which would again iterate over all the spans.

I tried to configure a translation like:

```erlang
 {translate_field,
   {[zipkin_list_of_spans, spans],
     [{encode, {opentelemetry_zipkin, e_fn, ['$1']}}]}}
```

But get an error:

```
{error, {missing_translation, {op,decode_init_default},
```

I don't see this mentioned in the translation docs https://github.com/tomas-abrahamsson/gpb/blob/c0fff60905f1a8f28c76ba47cceaaec5ed0d8a65/doc/dev-guide/translations.md

**Comment by tomas-abrahamsson** [comment]
It looks like it could fit well with what translations are aimed at. For fields in general, gpb will want at least a `decode` translator and an `encode` translator specified. Also even if you call the generated code only for encoding, you currently still will need to specify also a decode translator, but you can of course specify a dummy one if you expect it never to get called. It is also possible to specify `merge` and a `verify` translator, though these have defaults.

For repeated fields, gpb will also want `decode_init_default`, `decode_repeated_add_elem`, and `decode_repeated_finalize`. For example, if the internal format for a list would be a set, then `decode_init_default` would be to call the`sets:new/0` function, `decode_repeated_add_elem` would be `sets:add_element/2`, `decode_repeated_finalize` could be just an id function, and one might want to specify `merge` to call `sets:union/2`. The `encode` would call `sets:to_list/1`. Note that for a repeated field, these specified translators apply to the list-level itself, not to the elements. You probably want to specify translations to the `Span` elements, too; gpb will need to be able to see them too, on their normal format to encode them.

The many translations correspond to the different kind of situations where a repeated field can be handled. An octet stream is from a protobuf point of view only a stream of fields, For repeated fields, on decoding they can occur several times hence the need to add an item at a time, or not at all for an empty sequence, hence the need to specify an initial default value. The finalization stage can be utilized for amortizing the cost of adding an element at a time, for example for lists, `decode_repeated_add_elem` means to add a cons cell, and `decode_repeated_finalize` calls `lists:reverse/1`. For lists, the `merge` translator would normally call `lists:append/2` and would get called for example if the decoder would see two messages containing a repeated field in the same binary to decode.

(That guide in the dev-guide directory was mostly aimed for someone needing to dig into the gpb-internals of how translations are handled, but I realize there's more to write about it. I figure there might be a need to write some user doc too, about translations.)

**Comment by tsloughter** [comment]
>  Note that for a repeated field, these specified translators apply to the list-level itself, not to the elements.

Oh, so it isn't an encode function called for every Span in the repeated `spans` element?

Which, actually I just realized I don't even have a list at first, I have an ets table... so my original thinking won't work completely (though I can select all rows to a list of course).

Right now it does an `ets:fold` to convert each row record into the needed gpb record https://github.com/opentelemetry-beam/opentelemetry_zipkin/blob/master/src/opentelemetry_zipkin.erl#L26

**Comment by tomas-abrahamsson** [comment]
Hmm, you are probably right, if the translator for the repeated field would return a list of translated `spans` elements, it would probably work.

**Comment by tomas-abrahamsson** [comment]
Did you eventually get things to work? Would it be ok to close this ticket?

**Comment by tsloughter** [comment]
I didn't, but I also didn't spend much time on it yet. I'll close this though as the solution at least sounds like it is going to work eventually. Thanks!

**Comment by tomas-abrahamsson** [comment]
No problems, just reopen if there's anything more
